packages/testing/src/chai-retry-plugin/helpers.ts

```diff
@@ -0,0 +1,72 @@
+import Chai from 'chai';
+import { timeout as timeoutPromise, sleep } from 'promise-assist';
+
+import { chaiMethodsThatHandleFunction } from './constants';
+import type { RetryAndAssertArguments, RetryOptions } from './types';
+
+export const retryFunctionAndAssertions = async (retryAndAssertArguments: RetryAndAssertArguments): Promise<void> => {
+    let assertionError: Error | undefined;
+    let isTimeoutExceeded = false;
+
+    const performRetries = async ({ functionToRetry, options, assertionStack }: RetryAndAssertArguments) => {
+        const { retries, delay } = options;
+        let retriesCount = 0;
+
+        while (retriesCount < retries && !isTimeoutExceeded) {
+            try {
+                retriesCount++;
+                /**
+                 * If assertion chain includes such method as `change`, `decrease` or `increase` that means function passed to
+                 * the `expect` will be called by Chai itself
+                 */
+                const shouldAssertFunctionValue = assertionStack.some((stackItem) =>
+                    chaiMethodsThatHandleFunction.includes(stackItem.propertyName)
+                );
+                const valueToAssert = shouldAssertFunctionValue ? functionToRetry : await functionToRetry();
+                let assertion = Chai.expect(valueToAssert);
+
+                for (const { propertyName, method, args } of assertionStack) {
+                    if (method && args) {
+                        assertion = method.apply(assertion, args);
+                    } else {
+                        assertion = assertion[propertyName] as Chai.Assertion;
+                    }
+                }
+
+                return;
+            } catch (error: unknown) {
+                assertionError = error as Error;
+                await sleep(delay);
+            }
+        }
+
+        throw new Error(`Limit of ${retries} retries exceeded! AssertionError: ${assertionError}`);
+    };
+
+    const getTimeoutError = () => {
+        isTimeoutExceeded = true;
```

Comment:
this is wrong. In general don't change something not related in the `getSomething` function.
Besides this, here you don't even control when this function will be called (`timeoutPromise` might call it whenever it wants)

Reply:
@yurii-ve  Why? I know exactly when this function would be called - when timeout is exceeded, and I need to check that timeout isn't exceeded in `while` loop, what do we lose in such solution?

Reply:
@yurii-ve replaced with regular `setTimeout`